source/common/upstream/cds_api_impl.cc

```diff
@@ -30,11 +30,14 @@ CdsApiImpl::CdsApiImpl(const envoy::api::v2::core::ConfigSource& cds_config, Clu
     : cm_(cm), scope_(scope.createScope("cluster_manager.cds.")) {
   Config::Utility::checkLocalInfo("cds", local_info);
 
+  bool is_delta = (cds_config.api_config_source().api_type() ==
+                   envoy::api::v2::core::ApiConfigSource::DELTA_GRPC);
+  std::string grpc_method = is_delta ? "envoy.api.v2.ClusterDiscoveryService.DeltaClusters"
+                                     : "envoy.api.v2.ClusterDiscoveryService.StreamClusters";
   subscription_ =
       Config::SubscriptionFactory::subscriptionFromConfigSource<envoy::api::v2::Cluster>(
```

Comment:
Can you make `SubscriptionFactory` an injectable factory (rather than use a wrapper for a `static` method)? The main difficulty there might be the use of templates, but it could be possible to at least make the `Grpc::AsyncClientFactoryPtr` injectable..

Reply:
It would entail changes to everything that uses SubscriptionFactory, including all the different xDS _api_impls - I would prefer to keep it in a followup PR.

But, while we're discussing it: having looked into it, I agree that SubscriptionFactory is currently not as clean as it could be. But, rather than going further into factory-ness, I would actually argue for the opposite: get rid of SubscriptionFactory, and just have subscriptionFromConfigSource() as a standalone function. SubscriptionFactory is not currently a real factory, and rather is just a Java-style wrapper around what is more naturally a standalone function. As a result, nothing ever deals with a SubscriptionFactory object, and I don't see any way it would be beneficial.

Reply:
It would be beneficial in the same way any dependency injection is beneficial; you can create and inject mocks. Anyway, let's agree to do this later as we need to unblock end users, but please keep this on your agenda, as we should have tests.

Reply:
So, kind of centralizing what ClusterManagerFactory::createCds, ListenerComponentFactory::createLdsApi, etc currently do?

Reply:
I take that back, that wouldn't make sense. CdsApi / LdsApi are not sibling subclasses; they are different interfaces from include/. That's why tests are getting e.g. their MockCdsApis specifically from a mocked ClusterManagerFactory. Maybe there is some improvement to be made, but it seems like it would be a serious overhaul of the structure of the whole [xds]Api setup.